File: bms/go/geo/geo_s2_Cap_String/geo_s2_Cap_String.go
package test

import (
	"fmt"
	"math"
)

const (
	geo_s1_Radian geo_s1_Angle = 1
	geo_s1_Degree              = (math.Pi / 180) * geo_s1_Radian
)

func (c geo_s2_Cap) String() string {
	return fmt.Sprintf("[Center=%v, Radius=%f]", c.center.geo_r3_Vector, c.Radius().Degrees())
}

func (c geo_s2_Cap) Radius() geo_s1_Angle {
	return c.radius.Angle()
}

func (c geo_s1_ChordAngle) Angle() geo_s1_Angle {
	if c < 0 {
		return -1 * geo_s1_Radian
	}
	if c.IsInfinity() {
		return geo_s1_InfAngle()
	}
	return geo_s1_Angle(2 * math.Asin(0.5*math.Sqrt(float64(c))))
}

func (c geo_s1_ChordAngle) IsInfinity() bool {
	return math.IsInf(float64(c), 1)
}

func geo_s1_InfAngle() geo_s1_Angle {
	return geo_s1_Angle(math.Inf(1))
}

func (a geo_s1_Angle) Degrees() float64 { return float64(a / geo_s1_Degree) }

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
}

type geo_s1_ChordAngle float64

type geo_s1_Angle float64

File: translations_only/fail/go/geo/geo_s2_Cap_String/mistral/Restart/geo_s2_Cap_String.rs

use std::f64;
use std::fmt;

const GEO_S1_RADIAN: f64 = 1.0;
const GEO_S1_DEGREE: f64 = (std::f64::consts::PI / 180.0) * GEO_S1_RADIAN;

struct GeoS2Cap {
    center: GeoS2Point,
    radius: GeoS1ChordAngle,
}

struct GeoS2Point {
    geo_r3_vector: (),
}

struct GeoR3Vector {
}

struct GeoS1ChordAngle(f64);

struct GeoS1Angle(f64);

impl fmt::Display for GeoS2Cap {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "[Center={:?}, Radius={:?}]", self.center.geo_r3_vector, self.radius.0)
    }
}

impl GeoS2Cap {
    fn radius(&self) -> GeoS1Angle {
        GeoS1Angle(self.radius.0.to_radians())
    }
}

impl GeoS1ChordAngle {
    fn angle(&self) -> GeoS1Angle {
        if self.0 < 0.0 {
            return GeoS1Angle(-1.0 * GEO_S1_RADIAN);
        }
        if self.is_infinity() {
            return GeoS1Angle(f64::INFINITY);
        }
        GeoS1Angle(2.0 * (f64::asin(0.5 * (self.0).sqrt())))
    }

    fn is_infinity(&self) -> bool {
        self.0.is_infinite()
    }
}

impl GeoS1Angle {
    fn degrees(&self) -> f64 {
        self.0 / GEO_S1_DEGREE
    }
}

impl GeoS1ChordAngle {
    fn to_radians(&self) -> f64 {
        self.0 * (std::f64::consts::PI / 180.0)
    }
}

impl GeoS1Angle {
    fn to_radians(&self) -> f64 {
        self.0 * (std::f64::consts::PI / 180.0)
    }
}

fn geo_s1_inf_angle() -> GeoS1Angle {
    GeoS1Angle(f64::INFINITY)
}
